### ISSUE TITLE
Prisma lambda package size still seems large even after latest changes

### ISSUE THREAD

**Issue by imoby** [prisma/prisma#19358]
### Bug description

I've been following the serverless and lambda enhancements and performance improvements for quite some time that you all have made but I still seem to notice that my lambda even with the latest performance changes, still seems larger than expected.

I have deployed a serverless backend with nodejs using aws and my lambda right now currently sits at 25mbs zipped and almost 65mbs unpacked. This seems quite high still even after all the recent changes I expected it to have gone down quite a bit. I have verified that only the `libquery_engine-rhel-openssl-1.0.x.so.node` engine is whats packed up but there still seems to be a lot of other files/changes that add it up to be 65mbs. Curious if this is expected behavior.

The `libquery_engine-rhel-openssl-1.0.x.so.node` itself sits at 17mbs. I'm wondering whats causing so many of the other files to be so much. 

These two directories in the node_modules folder of my lambda already add up to quite a lot
![image](https://github.com/prisma/prisma/assets/6380855/8843a85d-bc9a-435a-a872-ebfb326738ed)

### How to reproduce

1. Setup a serverless lambda project
2. Use postgres prisma within it
3. Deploy changes to aws
4. View the size

### Expected behavior

Really shouldnt be this large of a size for prisma alone

### Prisma information

```prisma
// Add your schema.prisma
```

```ts
// Add your code using Prisma Client
```

Not applicable here

### Environment & setup

- OS: macOS
- Database: PostgreSQL
- Node.js version: 18

### Prisma Version

```
4.14.1
```

**Comment by SevInf** [comment]
Hi @imoby 
You don't need `prisma` package when deploying to lambda — this is a CLI that is needed only for development. In production, only `@prisma/client` is needed.

**Comment by imoby** [comment]
@SevInf can you confirm how I am suppose to exclude them because I dont have `prisma` as an package.json dependency, I just install it globally to use in the cli. But I am using serverless webpack prisma which builds my serverless project for me and when I look at the lambda output in aws and export it this is what the node modules looks like:

<img width="264" alt="image" src="https://github.com/prisma/prisma/assets/6380855/c433bc77-2b7e-450e-86b6-71f6864209f7">

Prisma dir:
<img width="328" alt="image" src="https://github.com/prisma/prisma/assets/6380855/8ccfed24-f717-4d98-9ee7-9c604f21211b">


@prisma dir
<img width="328" alt="image" src="https://github.com/prisma/prisma/assets/6380855/4bda0f97-c3c8-4d22-a762-33c21683357a">

I'm definitely missing something then

**Comment by janpio** [comment]
Can you share a minimal reproduction of your project? We only build Prisma, so to understand how it interacts with other software like "serverless webpack prisma" it really helps if we can look at actual code and play with it ourselves.

**Comment by imoby** [comment]
I can definitely try to work on getting a reproducible example at some point but essentially this is the contents of my specific lambda bundles package.json file

```
{
  "name": "sample",
  "version": "1.0.0",
  "description": "Packaged externals for sample",
  "private": true,
  "scripts": {},
  "dependencies": {
    "@prisma/client": "^4.14.1",
    "zod": "^3.17.3"
  }
}
```
[node_modules.zip](https://github.com/prisma/prisma/files/11560920/node_modules.zip)

With that, I am seeing the attached node_modules (I've compressed and attached it to this). I'm curious why it is such a large package and why my lambda is seeing such a high package size when my actual lambda function code itself is just 6kb

<img width="345" alt="image" src="https://github.com/prisma/prisma/assets/6380855/9035f423-6a89-4541-a089-1c78308c12e8">

**Comment by janpio** [comment]
Your `node_modules` of your function bundle contains these files because your Serverless Framework configuration makes it grab all these files - including Prisma CLI. Effectively that means it includes the Prisma Query Engine twice, which is not needed. It also includes two 4.5MB files to run the `prisma` CLI (`node_modules/.bin/prisma`) which is an unfortunate MacOS thing (On Windows these files are 2.82kb _in total_). They are also not needed..

We can not figure out via the end result _why_ that is, we need a way to understand what the third party framework is doing and how it decides it should include all these files.

Is the size of this bundle a problem for you? 
Does it make something slower than it should be?
(except for the upload, which of course will take a few ms longer when the bundle is a few megabytes bigger)

**Comment by imoby** [comment]
Size of the lambda should always be a consideration for both performance during cold starts and the possibility of reaching the limit of lambda size. In this case I'm more interested in optimizing for cold start but I can easily see someone running into the size upload limit for lambda so this should definitely be addressed somehow even if it means we manually remove some files like we do today with the additional engines that are not needed after build.

**Comment by janpio** [comment]
Of course, we fully agree. But is your cold start faster when you remove the superfluous files?

If you share your project setup so we can reproduce what you are seeing, we can understand why Serverless Framework includes all these dependencies insteaf of just the necessary ones. We can also suggest strategies to exclude those potentially.

**Comment by imoby** [comment]
It might actually help for me to better understand what files specifically are needed and are not. Based on the zipped node_modules I sent is it clear what should and should not be part of the build? This would be easier for me to not include them during the bundle and compare

**Comment by paulo2703** [comment]
I am facing the same issue. We are using (probably the same) a really basic setup using **serverless framework + serverless webpack + serverless webpack prisma**. While the unneccessary engine files in @prisma/client seem to perfectly be sorted out, the "prisma" library also gets uploaded, which (especially when individually bundling each function) **adds a 31.2mb per function** to upload. We have not been able to find a proper way how to sort it out, neither know we why the package is actually being bundled as we do not have it in our package.json, nor any other package used has it in its dependencies, except finding it as peer dependency for prisma-client.

Do you guys have a good guess or an idea why the "prisma" library could bundled?

**Comment by imoby** [comment]
Thank you @paulo2703 I'm glad we're not the only ones facing this issue because the bundle size really is way more than what our actual code is. I've even attempted to switch to other orm's just for comparison and I'm seeing drastic decrease in mb's

**Comment by paulo2703** [comment]
@imoby what would be interesting, the only other library related to prisma we use is prisma-nested-middleware. This one uses "prisma" as devDependency. Might this be the cause? Do you also use it?

**Comment by imoby** [comment]
we dont use that library at all no @paulo2703

**Comment by janpio** [comment]
Yes, that could certainly explain what you are seeing @paulo2703. Can you open a new issue and provide some more information about your project, optimally a reproduction of your setup? I bet there is a way that we can use to configure your project to also ignore that dependency. Thanks.

**Comment by paulo2703** [comment]
@imoby nevermind. I still have the issue when I cut out everything related to `prisma-nested-middleware`. 

After investigating further, I found out that the webpack bundler (please confirm that I'm not wrong?) also includes peerDependencies when bundling. The "prisma" library is actually a peerDependency of @prisma/client, therefore it might be getting bundled. Unfortunately I still couldn't figure out how to exclude it from the bundling.

**Comment by janpio** [comment]
Can you provide a reproduction of this behavior @paulo2703? That would be optimal.
@imoby Are you also using Webpack?

**Comment by imoby** [comment]
@janpio Yes I am using the same serverless webpack option that is outlined here https://www.prisma.io/docs/guides/deployment/deployment-guides/use-prisma-with-serverless-framework#deployment-with-serverless-webpack

**Comment by imoby** [comment]
Also, just noticed this too :) https://github.com/danieluhm2004/serverless-webpack-prisma/issues/22

**Comment by paulo2703** [comment]
@janpio Was fun destroying the whole project once ;) a.k.a. Please find a reproduction here: https://github.com/paulo2703/webpack-prisma-reproduction

just give it a npm run package and look into the generated zip file in .serverless. It containes the prisma folder although not mentioned anywhere beside as a peerDependency in the client.

**Comment by paulo2703** [comment]
@imoby I am trying to reproduce the installDeps: false in the repo but the size seems to be identical to what it was before. Could you try this also at your end?

**Comment by janpio** [comment]
@paulo2703 With your reproduction I can definitely already confirm that your package includes the same `libquery_engine-rhel-openssl-1.0.x.so.node` twice, and also the CLI `.bin` files which weigh 2x 4.2MB on Linux for some weird Npm reason. Getting rid of the CLI before the packaging runs, should remove both teh second Query Engine copy and both the `.bin` files.

**Comment by paulo2703** [comment]
@janpio how would you practically do this? Can you provide a code example?

**Comment by janpio** [comment]
Not instantly, as this fully depends on your project and Serverless Framework setup. 

If I was in your position, I would start with a fake project that I newly create, and then I add and enable things one by one and track the size. You should be able to start with a package that really only includes the needed Prisma Client and its engine once. Then when you recreate what you have in your real app step by step, at a certain point it will suddenly include Prisma CLI as well - so you know what causes the problem.

**Comment by janpio** [comment]
Your core problem is that your package includes `prisma`, the Prisma CLI. You don't need and want the CLI on Lambda. If you remove that, both the second engine files and the 2 `.bin` files are not present.

**Comment by paulo2703** [comment]
@janpio Yes as mentioned I understood that but the way on how to get rid of is is not fully clear. As it is a peerDep of @prisma/client (I never explicitly import or installed prisma in the reproduction.) it will always be included into the package if I include @prisma/client. The question now is how to overcome this. Right?

**Comment by janpio** [comment]
Yes, or how to make your tooling understand that you don't want it. In my recollection with Serverless Framework we usually suggested to install `prisma` as a devDependency, as those were excluded by default - but I am not sure how that connects to what you are describing. (That might just have been in the non Webpack use cases - not sure any more)

**Comment by janpio** [comment]
Hey, any change here @luap2703 and @imoby? We basically do not hear this complaint any more from other users - which was _very_ different a few years ago before we optimized and fixed this - so I am wondering if this might also be resolved for you? Or maybe we can help you figure out the configuration required to optimize your specific situation (and potentially document it in our docs as well, for others)?

**Comment by imoby** [comment]
I hate to sound negative but I still think the size is a big issue. I love prisma for many things and I know you all are working towards improving a lot of support for better drivers on things like these third party serverless services but IMO its hard to justify 13-15 mbs of lambda and difficult cold start. I do agree it has gotten much better comparing before and after the major upgrade you all did but I still see a use case for a very very slim version of your engine just for serverless instances and I dont see any discussions around that.

**Comment by janpio** [comment]
I understand, that is why I am asking - so no need to be cautious. We will not take it as negative as long as the discussion stays productive 🤗 

Do you have concrete measurements how this negatively influences cold start of your functions? I know the general advice on the internet is "bigger Lambda = slow cold start", but we are not aware of _concrete_ measurements showing a problem with our current size. We think that is because 99% of our size is from one single file (the query engine) - while the typical benchmarks are often done with JavaScript projects that have tens of thousands of files. 
(Benchmarks of Prisma Client in Lambda environments also often includes the opening of the database connection in the "cold start" measurement, which misattributes the time required to open the TCP connection to "Prisma Client makes cold start slow" but would also exist with any other database client or ORM.)

**Comment by luap2703** [comment]
Thank you for catching up Jan!

So we resolved this issue by moving Prisma into an aws lambda layer as **not the cold-start time was our main concern, but rather the time to deploy**. Think of 20 different lambdas pointing to the same Prisma module and all having to upload the query engine(s) separately. Having also uploaded all (in this case) unneccessary other engine files was just blowing up the deployment to 500mb-2gb aka sometimes up to 30min, just because of the duplicated engine files.

We now implemented stack-based **prisma layers** (saying we have 3 microservices (stacks) with different Prisma schemes and dbs, then there'll be 3 layers deployed) so the engine only has to be uploaded once.

Additionally, we **remove all unnecessary engines during bundling**, to make sure only the right one gets deployed.

Maybe as a suggestion for future users, it might be helpful to point towards such layers in the docs.

Here is our aws cdk implementation of the prisma layer for anyone curious:`

```typescript
import { AssetHashType, IgnoreMode, SymlinkFollowMode } from "aws-cdk-lib";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { Code, Runtime } from "aws-cdk-lib/aws-lambda";
import { Construct } from "constructs";
import * as crypto from "crypto";
import * as fs from "fs";

export interface PrismaLayerProps
    extends Omit<lambda.LayerVersionProps, "code"> {
    prismaRoot: string;

    // e.g. 3.1.1
    prismaVersion?: string;

    // some more modules to add to the layer
    nodeModules?: string[];
}

export class PrismaLayer extends lambda.LayerVersion {
    readonly extModules: string[];
    readonly environment: Record<string, string>;
    readonly prismaRoot: string;

    constructor(scope: Construct, id: string, props: PrismaLayerProps) {
        const { prismaVersion, prismaRoot } = props;
        const nodeModules = props.nodeModules || [];

        const PRISMA_DEPS = ["prisma", "@prisma/migrate", "@prisma/client"];
        let modulesToInstall = PRISMA_DEPS.concat(nodeModules);
        if (prismaVersion)
            modulesToInstall = modulesToInstall.map(
                (dep) => `${dep}@${prismaVersion}`,
            );
        const modulesToInstallArgs = modulesToInstall.join(" ");

        const assetOut = "/asset-output/nodejs";

// Remove all non-required engines here 
        const createBundleCommand = [
            "bash",
            "-c",
            [
                `mkdir -p ${assetOut}`,
                `cp -R . ${assetOut}/prisma`,
                `cd ${assetOut}`,
                `export PRISMA_CLI_BINARY_TARGETS=rhel-openssl-1.0.x`,
                `HOME=/tmp npm i --omit=dev ${modulesToInstallArgs}`,
                `HOME=/tmp npx prisma generate --generator client`,
                `rm -rf ${assetOut}/node_modules/.cache`,
                `rm -rf ${assetOut}/node_modules/.prisma/client/*linux* || true`,
                `rm -rf ${assetOut}/node_modules/@prisma/engines/*linux* || true`,
                `rm -rf ${assetOut}/node_modules/prisma/*engine* || true`,
            ].join(" && "),
        ];

        const prismaSchema = fs.readFileSync(`${prismaRoot}/schema.prisma`);
        const bundleCommandHash = crypto.createHash("sha256");
        bundleCommandHash.update(JSON.stringify(props));
        bundleCommandHash.update(prismaSchema);

        const code = Code.fromAsset(prismaRoot, {
            assetHash: bundleCommandHash.digest("hex"),
            assetHashType: AssetHashType.CUSTOM,
            bundling: {
                image: Runtime.NODEJS_20_X.bundlingImage,
                command: createBundleCommand,
            },
            followSymlinks: SymlinkFollowMode.ALWAYS,
            ignoreMode: IgnoreMode.GIT,
        });

        super(scope, id, { ...props, code });

        this.prismaRoot = prismaRoot;

        // hint for prisma to find the engine
        this.environment = {
            PRISMA_QUERY_ENGINE_LIBRARY:
                "/opt/nodejs/node_modules/.prisma/client/libquery_engine-rhel-openssl-1.0.x.so.node",
            PRISMA_SCHEMA_ENGINE_LIBRARY:
                "/opt/nodejs/node_modules/@prisma/engines/schema-engine-rhel-openssl-1.0.x",
        };

        // modules provided by layer
        this.extModules = [
            "@aws-sdk/*",
            "@prisma/client",
            "@prisma/migrate",
            "@prisma/sdk",
            "@prisma/engines",
            "@prisma/engines-version",
            "@prisma/prisma-schema-wasm",
            ...nodeModules,
        ];
    }
}

```

**Comment by janpio** [comment]
That makes a _lot_ of sense, and an aspect that is not brought up that often and hence often skipped. Smaller engine files would of course make this problem less pressing, but you seem to have found a good solution. Do you think your layer definition would make a good case for a example aws cdk implementation?

**Comment by ilyaLibin** [comment]
@luap2703 Can you please share what package size of the lambda you was able to get after moving prisma to the layer?
Mine are 12.8mb and still seems way to big and include a lot of unrelated code.
